05-assistive-chatbot/chatbot/engines/v2_household_engine.py

```diff
@@ -1,56 +1,194 @@
+
+    @utils.timer
+    def gen_response(self, query):
+        derived_questions = decompose_with_retries(self.decomposer, query)
+        derived_question_entries = retrieve_cards_for(
+            derived_questions, self.vectordb_wrapper.vectordb, self.retrieve_k
+        )
+        card_entries = collate_by_card_score_sum(derived_question_entries)
+
+        gen_results = GenerationResults(query, derived_question_entries, card_entries)
+        populate_summaries(gen_results, self.guru_card_texts, self.summarizer)
+        logger.debug("gen_results: %s", gen_results)
+        return gen_results
+
+    def decomposer(self, query):
+        response = self.call_llm(self.decomposer_client, self.prompts.decomposer, question=query)
+        return json.loads(response)
+
+    def summarizer(self, **kwargs):
+        return self.call_llm(self.summarizer_client, self.prompts.summarizer, **kwargs)
+
+    def call_llm(self, llm_client, prompt_template: str, **template_inputs):
+        prompt = prompt_template.format(**template_inputs)
+        logger.debug("Prompt: %s", prompt)
+        response = llm_client.generate_reponse(prompt)
+        logger.debug("Response object: %s", response)
+        return response
 
-        context_question = dspy.InputField()
-        context = dspy.InputField()
-        answer = dspy.OutputField()
 
-    return dspy.Predict(SummarizeCardGivenQuestion)
+def decompose_with_retries(decomposer, query):
+    for i in range(1, 4):  # retry loop
+        if i > 1:
+            logger.warning("Retrying to get parsable JSON response -- attempt %i", i)
+            # TODO: also send notification to UI by adding a message to GenerationResults
+        try:
+            derived_questions = decomposer(query)
+
+            logger.info("Derived questions: %s", derived_questions)
+            return derived_questions
+        except json.JSONDecodeError as e:
+            logger.error("Error decomposing question: %s", e)
+    return []
+
+
+## Non-LLM Helper functions
+
+
+def collate_by_card_score_sum(derived_question_entries):
+    all_retrieved_cards = dict()
+    for dq_entry in derived_question_entries:
+        scores = dq_entry.retrieval_scores
+        for i, card in enumerate(dq_entry.retrieved_cards):
+            all_retrieved_cards[card] = all_retrieved_cards.get(card, 0) + scores[i]
+
+    card_to_dqs = {}
+    card_to_quotes = {}
+    for dq_entry in derived_question_entries:
+        derived_question = dq_entry.derived_question
+        for card in dq_entry.retrieved_cards:
+            if card not in card_to_dqs:
+                card_to_dqs[card] = set()
+            card_to_dqs[card].add(derived_question)
+
+        for card, quote in zip(dq_entry.retrieved_cards, dq_entry.retrieved_chunks):
+            if card not in card_to_quotes:
+                card_to_quotes[card] = set()
+            if quote != card:
+                card_to_quotes[card].add(quote)
```

Comment:
If the quote == card title, then don't use it as a quote b/c the card title will already be shown in the UI and the title/question is not particular helpful as a quote.